#!/usr/bin/env python
#
# Authors: Gregory S Mendez and Bastian Bentlage
#
# This script fetches sequences listed in blast results files and writes out a plain text
# file to be used by another script to write new fasta files with the full length sequences
# using the def-lines listed in the blast results and a large fasta file used to generate
# the blastdb the blast results came from.
#
# This script takes 2 arguments:
# 1) --blast - The directory containing the blast output files (in XML format)
# 2) --outdir - The directory to write the text files to
#
# Usage: PepFromBlast.py --blast ~/GreenAlgae/big_blastp/ --outdir ~/GreenAlgae/pepfromblast_txt/
#
# The script get_seq.sh should be used after this script to write the fasta files using the text
# files generated by this script. 

from Bio.Blast import NCBIXML
from glob import glob
from Bio import SeqIO, Seq
import sys
import argparse
#from pyfaidx import Fasta

def ExtractPeps(BLAST_Results, OutDirectory):
	SPECIES_GENE = {}
  # Loop over the BLAST XML and retrieve the compIDs
	for BLASTout in glob('%s*out' % BLAST_Results):
		# Define species and gene ID
		SPECIES = '_'.join(BLASTout.split('/')[-1].split('.')[1].split('_')[2:])
		GENE = BLASTout.split('query_')[-1].split('.')[0]
		# Create a list of ORFs we want to retrieve from Transdecoder
		# Take the sequence ID from the BLAST XML and put it into a list
		for blast_record in NCBIXML.parse(open(BLASTout)):
			for Alignment in blast_record.alignments:
				ORF_ID = str(Alignment).split()[1]
				try:
					SPECIES_GENE[SPECIES][GENE].add(ORF_ID)
				except KeyError:
					try:
						SPECIES_GENE[SPECIES]
						SPECIES_GENE[SPECIES][GENE] = set()
						SPECIES_GENE[SPECIES][GENE].add(ORF_ID)
					except KeyError:
						SPECIES_GENE[SPECIES] = {}
						SPECIES_GENE[SPECIES][GENE] = set()
						SPECIES_GENE[SPECIES][GENE].add(ORF_ID)
				continue
# Write text files for each gene_species of all the sequences to be fetched by a 
# separate script: get_seq.sh
	for SPECIES, GENE_ORFS in SPECIES_GENE.iteritems():
		for KOG, GENE_ORF in GENE_ORFS.iteritems():
			OutFasta = '%s/%s_%s.txt' % (OutDirectory,KOG, SPECIES)
			with open(OutFasta, 'a') as Out:
				for VALUE in list(GENE_ORF):
					Out.write('%s\n' % VALUE)
					
# Alternative code to write fasta files out directly rather than just text files.
# These methods are much much slower than outputting text files then fetching the
# sequences using the separate script: get_seq.sh which is paralleled.
#
# pyfaidx version
#  	for SPECIES, GENE_ORFS in SPECIES_GENE.iteritems():
#  		for KOG, GENE_ORF in GENE_ORFS.iteritems():
# 			ORFs = '%s%s%s' % (ORF_Directory, SPECIES, '.fa.transdecoder.pep')
# 			GENES = Fasta(ORFs)
# 			OutFasta = '%s/%s_%s.faa' % (OutDirectory,KOG, SPECIES)
# 			with open(OutFasta, 'a') as Out:
# 				for KEY in list(GENE_ORF):
# 					Out.write('>%s-%s\n%s\n' % (SPECIES, GENES[KEY].name, GENES[KEY]))
					
# biopython version
# 	for SPECIES, GENE_ORFS in SPECIES_GENE.iteritems():
# 		# Where are the ORFs located - transdecoder
# 		ORFs = '%s%s%s' % (ORF_Directory, SPECIES, '.fa.transdecoder.pep')
# 		with open(ORFs, 'rU') as FastaFile:
# 			for record in SeqIO.parse(FastaFile, "fasta"):
# 				for KOG, GENE_ORF in GENE_ORFS.iteritems():
# 					for ORF in GENE_ORF:
# 						if record.name == ORF:
# 							OutFasta = '%s/%s_%s.faa' % (OutDirectory,KOG, SPECIES)
# 							with open(OutFasta, 'a') as Out:
# 								Out.write('>%s-%s\n%s\n' % (SPECIES, record.name, record.seq))

# Argument Parser
parser = argparse.ArgumentParser(description = 'This script fetches sequences listed in blast results files and writes out a plain text file to be used by another script to write new fasta files with the full length sequences using the def-lines listed in the blast results and a large fasta file used to generate the blastdb the blast results came from.')
parser.add_argument('--blast', required=True, help='BLAST results XML directory') 
parser.add_argument('--outdir', required=True, help='Output gets written here') 
args = parser.parse_args() 

ExtractPeps(args.blast, args.outdir)
